packages/user/TokenStream/service/src/tests.rs

```diff
@@ -0,0 +1,396 @@
+        TokenStream::push(&chain)
+            .get_stream(nft_id)
+            .get()
+            .unwrap()
+            .unwrap()
+    }
+
+    fn tokens_credit(
+        chain: &psibase::Chain,
+        token_id: u32,
+        creditor: AccountNumber,
+        debitor: AccountNumber,
+        amount: u64,
+    ) {
+        Tokens::push_from(&chain, creditor)
+            .credit(
+                token_id,
+                debitor,
+                amount.into(),
+                "memo".to_string().try_into().unwrap(),
+            )
+            .get()
+            .unwrap();
+    }
+
+    fn setup_env(chain: &psibase::Chain) -> Result<u32, psibase::Error> {
+        chain.new_account(ALICE)?;
+        chain.new_account(BOB)?;
+        chain.new_account(CHARLIE)?;
+
+        let supply = Quantity::from(10000000);
+
+        let token_id = Tokens::push_from(&chain, ALICE)
+            .create(4.try_into().unwrap(), supply)
+            .get()
+            .unwrap();
+        assert_eq!(token_id, 2);
+
+        Tokens::push_from(&chain, ALICE).mint(
+            token_id,
+            supply,
+            "memo".to_string().try_into().unwrap(),
+        );
+        Ok(token_id)
+    }
+
+    fn reset_clock(chain: &psibase::Chain) {
+        let time: i64 = 1000;
+        chain.start_block_at(TimePointSec { seconds: time }.microseconds());
+    }
+
+    fn create_stream(
+        chain: &psibase::Chain,
+        author: AccountNumber,
+        decay_rate: u32,
+        token_id: u32,
+    ) -> u32 {
+        TokenStream::push_from(&chain, author)
+            .create(decay_rate, token_id)
+            .get()
+            .unwrap()
+    }
+
+    #[psibase::test_case(packages("TokenStream"))]
+    fn test_basics(chain: psibase::Chain) -> Result<(), psibase::Error> {
+        chain.start_block();
+
+        reset_clock(&chain);
+
+        let token_id = setup_env(&chain)?;
+
+        tokens_credit(&chain, token_id, ALICE, TOKEN_STREAM, 500);
+
+        let stream_nft_id = create_stream(&chain, ALICE, 10000, token_id);
+
+        check_balance(&chain, token_id, TOKEN_STREAM, Some(0));
+
+        assert_eq!(
+            get_shared_balance(&chain, token_id, ALICE, TOKEN_STREAM),
+            500.into()
+        );
+
+        TokenStream::push_from(&chain, ALICE)
+            .deposit(stream_nft_id)
+            .get()
+            .unwrap();
+
+        assert_eq!(
+            get_stream(&chain, stream_nft_id).total_deposited,
+            500.into()
+        );
+
+        Nfts::push_from(&chain, ALICE).credit(stream_nft_id, BOB, "memo".to_string());
+        Nfts::push_from(&chain, BOB).debit(stream_nft_id, "memo".to_string());
+
+        TokenStream::push_from(&chain, BOB)
+            .claim(stream_nft_id)
+            .get()
+            .unwrap();
+
+        check_balance(&chain, token_id, BOB, Some(1));
+
+        chain.start_block_at(
+            TimePointSec {
+                seconds: 3600 * 7 * 52 * 20,
```

Comment:
What is this supposed to represent? We should be able to calculate the time at which the remaining unvested balance flushes to zero.

Reply:
02ea0d01faef7cdef9d0c37a1bacd2525ff7044c

Reply:
02ea0d01faef7cdef9d0c37a1bacd2525ff7044c